File: Assignment 2/yield_fert_rain-by Aneela Tabassum 2017-ag-3643 .R
#########################################
### """  Assignment # 2 : MS(CS) - Y | X1 | X2 .......Stats 701 :
Student name: Aneela Tabassum ,
AG no. 2017-ag-3643 ,
Degree and Session : MS(CS) 2017-2019 """

# Note: Firstly,I just calculation all this calculation on regresster then write this one code.
# The graphs code was taken form class mate "Arsalan Ali"

#########################################

#Installing packages for 3D-graph
install.packages("scatterplot3d") 
library("scatterplot3d") 

#########################################
###			 DATA
#########################################

Fertilizer = c(100, 200, 300, 400, 500, 600, 700)   
Rainfall = c(10, 20, 10, 30, 20, 20, 30) 
Yield = c(40, 50, 50, 70, 65, 65, 80)  
#Making Data-frame
df = data.frame(Fertilizer,Rainfall,Yield)

#########################################
###			3D GRAPH CODE
#########################################

# Code for the 3D-plot 
data(df)
head(df)
#scatterplot3d(x=Fertilizer, y=Rainfall, z=Yield)
s3d <- scatterplot3d(df, type = "h", color = "blue",
    angle=55, pch = 16,grid=TRUE, box=FALSE,


              main="Multiple Linear Regression \n 
               By Aneela Tabassum\n 2017-ag-3643",
              xlab = "Fertilizer (Pound)",
              ylab = "Rainfall (Inches)",
              zlab = "Yield (Bushel)")

# Using Pre-defined MLR function for the 3D-graph
my.lm <- lm(Yield~Fertilizer+Rainfall)
s3d$plane3d(my.lm)


#############################################



###				FUNCTION



#############################################

#This is the Custom MLR function
MLR<- function( X1, X2, Y )
{

	# Size of the input, 
	n = length(Y)

	# Mean of X1, X2 and Y
	mean.X1 = mean(X1)
	mean.X2 = mean(X2)
	mean.Y = mean(Y)


	x1 = X1 - mean.X1
	x2 = X2 - mean.X2
	y = Y - mean.Y
	

	Sum_x1 = sum(x1)
	Sum_x2 = sum(x2)
	Sum_y = sum(y)

		x1_square = sum((x1)^2)
	x2_square = sum((x2)^2)
	y_square = sum((y)^2)

		Sum_x1y = sum((x1*y))
	Sum_x1_square = sum((x1)^2)
	Sum_x2_square = sum((x2)^2)
	Sum_x2y = sum((x2*y))
	Sum_x1x2 = sum((x1*x2))

	Sum_x1y_Sum_x2_square = Sum_x1y*Sum_x2_square
	Sum_x2y_Sum_x1x2 = Sum_x2y*Sum_x1x2
	Sum_x1_square_Sum_x2_square = Sum_x1_square*Sum_x2_square
	Sum_x1x2_square = (Sum_x1x2)^2

	Beta1 = (Sum_x1y_Sum_x2_square - Sum_x2y_Sum_x1x2) / (Sum_x1_square_Sum_x2_square - Sum_x1x2_square)

	Beta2 = ((Sum_x2y*Sum_x1_square) - (Sum_x1y*Sum_x1x2)) / (Sum_x1_square_Sum_x2_square - Sum_x1x2_square)

	Beta0 = mean.Y - (Beta1*mean.X1) - (Beta2*mean.X2)

	# Number of parameters e.g, All the Betas
	k = 3	

	#  Y Hat: The Y-predicted
	Y_hat = Beta0 + Beta1*Fertilizer + Beta2*Rainfall
	Sum_Y_hat = sum(Beta0 + Beta1*Fertilizer + Beta2*Rainfall)


	TSS = sum((Yield - mean.Y)^2)
	MSS = sum((Y_hat - mean.Y)^2)
	RSS = sum((Yield - Y_hat)^2)

	# The one and only Regression Co-efficient
	R_square = MSS / TSS

	# Mean Square Error
	MSE = RSS / (n - k)

	# Beta1
	V_Beta1 = MSE*((Sum_x2_square) / (Sum_x1_square_Sum_x2_square - Sum_x1x2_square))
	SE_Beta1 = sqrt(V_Beta1)

	# Beta2
	V_Beta2 = MSE*((Sum_x1_square) / (Sum_x1_square_Sum_x2_square - Sum_x1x2_square))
	SE_Beta2 = sqrt(V_Beta2)

	#  Beta0
	mean.X1_square = mean.X1^2
	mean.X2_square = mean.X2^2
	mean.X1_mean.X2 = mean.X1*mean.X2

	# Biggest secret of all from Beta0's side
	V_Beta0 = MSE*((1 / n) + ((mean.X1_square*Sum_x2_square + mean.X2_square*Sum_x1_square - 2*mean.X1_mean.X2*(Sum_x1x2)) / (Sum_x1_square_Sum_x2_square - Sum_x1x2_square)))
	SE_Beta0 = sqrt(V_Beta0)

	# These are all the Outputs:

	cat("Mean of X1 = ", mean.X1, "\n")
	cat("Mean of X2 = ", mean.X2, "\n")
	cat("Mean of Y = ", mean.Y, "\n")
	cat("\n")
	cat("Sum of Deviations of X1 = " ,Sum_x1, "\n")
	cat("Sum of Deviations of X2 = " ,Sum_x2, "\n")
	cat("Sum of Deviations of Y = " ,Sum_y, "\n")
	cat("\n")
	cat("Sum of Square of X1 : Sum_x1 = " ,x1_square, "\n")
	cat("Sum of Square of X2 : Sum_x2 = " ,x2_square, "\n")
	cat("Sum of Square of Y : Sum_y = " ,y_square, "\n")
	cat("\n")
	cat("Beta1  = " ,Beta1, "\n")
	cat("Beta2  = " ,Beta2, "\n")
	cat("Beta0  = " ,Beta0, "\n")
	cat("\n")
	cat("Sum of Y-hat = ", Sum_Y_hat , "\n")
	cat("\n")
	cat("Total Sum of Squares : TSS = ", TSS , "\n")
	cat("Model Sum of Squares : MSS = ", MSS , "\n")	
	cat("Residual Sum of Squares : RSS = ", RSS , "\n")
	cat("\n")
	cat("Regression Co-efficient : R^2 = ", R_square, "\n")
	cat("\n")
	cat("Mean Square Error : MSE = ",MSE,"\n")
	cat("\n")
	cat("Variance of Beta1 = ",V_Beta1, "\n")
	cat("Standard Error of Beta1 = ",SE_Beta1, "\n")
	cat("\n")
	cat("Variance of Beta2 = ",V_Beta2,"\n")
	cat("Standard Error of Beta2 = ",SE_Beta2, "\n")
	cat("\n")
	cat("Variance of Beta0 = ",V_Beta0,"\n")
	cat("Standard Error of Beta0 = ",SE_Beta0, "\n")

	df2 = data.frame(X1,X2,Y)

	# Since i can't do Anova etc manually so here's the pre-defined ones
	fit <- lm(Y ~ X1 + X2, data=df2)
	summary(fit)

	}

# Combines the Data-set of Fertilizer , Rainfall and Yield
Exp.data = cbind(Fertilizer , Rainfall , Yield)


colnames(Exp.data) = c('Fertilizer' , 'Rainfall' , 'Yield')

# Calls the Table-Data function to generate Table
Exp.data

# This will load the entire Regression-Calculation process

Exp.regression<- MLR( Fertilizer, Rainfall, Yield )

# Executes the above function
Exp.regression

File: Assignment 2/yield_fert_rain-by Aneela Tabassum 2017-ag-3643 .py
"""  Assignment # 2 : MS(CS) - Yield | Fertilizer | Rainfall .......Stats 701 :
Student name: Aneela Tabassum ,
AG no. 2017-ag-3643 ,
Degree and Session : MS(CS) 2017-2019 """

# Note: Firstly,I just calculation all this calculation on regresster then write this one code.
# The graphs code was taken form class mate "Arsalan Ali"


# mpl_toolkits is for the actual 3D-plotting of our data-set , it's so cool,  it looks gray but it works
from mpl_toolkits.mplot3d import Axes3D

# It's for the normal graph plotting we do , nothi'n too special
from matplotlib import pyplot as plt

# My man numpy sure does all the stats
import numpy as np

import pandas as pd

# You can't have Science without Maths right? Just added it since i had to use Positive-Square Root for SE
import math as math

# Consider it a shortcut for Stats ( Pre-defined methods ), which i only applied in the end for the 3D graph
from statsmodels.formula.api import ols

# Anova Vanova, what would be the world without ya?
from statsmodels.stats.anova import anova_lm

from scipy import stats







#these are all data set:
Yield = np.array([40 , 50 , 50 , 70 , 65 , 65 , 80 ])
Fertilizer = np.array([100,200,300,400,500,600,700])
Rainfall = np.array([10, 20,10,30,20,20,30])

# Number of Observations
n = Rainfall.size
print("\nNumber of Observations : n = ", n)

# Calculate the mean of Yield , Fertilizer , Rainfall:
Ybar = sum(Yield) / Yield.size
Fbar = sum(Fertilizer) / Fertilizer.size
Rbar = sum(Rainfall) / Rainfall.size

print("\nYield Mean : Y = ", Ybar)
print("Fertilizer Mean : X1 = ", Fbar)
print("Rainfall Mean : X2 = ", Rbar)

# Let's deviate from the mean :
x1 = Fertilizer - Fbar
x2 = Rainfall - Rbar
y = Yield - Ybar

print("\nx1 =", x1)
print("x2 =", x2)
print("Y =", y)

# Let's take the sum of X1,X2 and Y  deviate :
Σx1 = np.sum(x1)
Σx2 = np.sum(x2)
Σy = np.sum(y)

# taking the sequre of  sums of X1,X2 and y's deviations :
square_x1 = np.sum((x1) ** 2)
square_x2 = np.sum((x2) ** 2)
square_Y = np.sum((y) ** 2)

# Just some formulas in order to get all the Betas
Σx1y = np.sum((x1 * y))
Σx1_square = np.sum((x1) ** 2)
Σx2_square = np.sum((x2) ** 2)
Σx2y = np.sum((x2 * y))
Σx1x2 = np.sum((x1 * x2))

Σx1y_Σx2_square = Σx1y * Σx2_square
Σx2y_Σx1x2 = Σx2y * Σx1x2
Σx1_square_Σx2_square = Σx1_square * Σx2_square
Σx1x2_square = (Σx1x2) ** 2
temp1 = Σx1_square_Σx2_square - Σx1x2_square

Beta1 = (Σx1y_Σx2_square - Σx2y_Σx1x2) / temp1
print("\nBeta1 = ", Beta1)

Beta2 = ((Σx2y * Σx1_square) - (Σx1y * Σx1x2)) / temp1
print("Beta2 = ", Beta2)

Beta0 = Ybar - (Beta1 * Fbar) - (Beta2 * Rbar)
print("Beta0 = ", Beta0)

# K is the Number of parameters
k = 3
print("\nNumber of Parameters from  B0, B1, B2 : k = ", k)

# Y hat is  The Y-predicted :
Y_hat = Beta0 + Beta1 * Fertilizer + Beta2 * Rainfall
ΣY_hat = np.sum(Beta0 + Beta1 * Fertilizer + Beta2 * Rainfall)
print("\nSum of All Y-hat values = ", ΣY_hat)

print("\nY = {} + {}X1 + {}X2".format(Beta0, Beta1, Beta2))

#calculate the TSS , MSS and RSS for computing the Regression Co-efficient
TSS = np.sum((Yield - Ybar) ** 2)
MSS = np.sum((Y_hat - Ybar) ** 2)
RSS = np.sum((Yield - Y_hat) ** 2)

print("\nTotal Sum of Squares : TSS = ", TSS)
print("Model Sum of Squares : MSS = ", MSS)
print("Residual Sum of Squares : RSS = ", RSS)

# Regression Co-efficient
R_square = MSS / TSS
print("\nR_square = ", R_square)

# Mean Square Error
MSE = RSS / (n - k)
print("\nMeans Square Error : MSE = ", MSE)

# Beta1's variances
temp2 = Σx1_square_Σx2_square - Σx1x2_square
V_Beta1 = MSE * ((Σx2_square) / temp2)
SE_Beta1 = math.sqrt(V_Beta1)
print("\nVariance of Beta1 = ", V_Beta1)
print("Standard Error of Beta1 : SE(B1) = ", SE_Beta1)

# Beta2's variances
V_Beta2 = MSE * ((Σx1_square) / temp2)
SE_Beta2 = math.sqrt(V_Beta2)
print("\nVariance of Beta2 = ", V_Beta2)
print("Standard Error of Beta2 : SE(B2) = ", SE_Beta2)


F_bar_square = Fbar ** 2
R_bar_square = Rbar ** 2
F_bar_R_bar = Fbar * Rbar

#  Beta0's variances
V_Beta0 = MSE * ((1 / n) + ((F_bar_square * Σx2_square + R_bar_square * Σx1_square - 2 * F_bar_R_bar * (Σx1x2)) / temp2))
SE_Beta0 = math.sqrt(V_Beta0)
print("\nVariance of Beta0 = ", V_Beta0)
print("Standard Error of Beta0 : SE(B0) = ", SE_Beta0)

# now goning to draw the 3'D graph
data_frame = pd.DataFrame(
    {
        "Yield": Yield
        , "Fertilizer": Fertilizer
        , "Rainfall": Rainfall

    }
)

# It will turn it into a 3D Graph
ax = plt.figure().gca(projection='3d')

# It will take the values of Y,X1,X2 and turn them into X,Y,Z Axis
ax.scatter(
    data_frame["Fertilizer"]
    , data_frame["Rainfall"]
    , data_frame["Yield"]
    , color="blue"
    , marker="o"
    , alpha=1
)


# From here....graph code is taken from class mate "Arsalan Ali"

Reg = ols(formula="Yield ~ Fertilizer + Rainfall", data=data_frame)

Fit2 = Reg.fit()
print("\n", Fit2.summary())
print("\n", anova_lm(Fit2))

# Again plotting our Dear 3D-Graph
ax = plt.figure().gca(projection='3d')

# Creating Axis X,Y,Z out of our Data Y,X1,X2
ax.scatter(
    data_frame["Fertilizer"]
    , data_frame["Rainfall"]
    , data_frame["Yield"]
    , color="blue"
    , marker="o"
    , alpha=1
)

# Title of the graph
ax.set_title('Regression Shape in 3D \n ''Aneela Tabassum 2017-ag-3643')

# Again labelling the Axis we just created , writing duplicate code is a/an headache
ax.set_xlabel("Fertilizer ", color="red")
ax.set_ylabel("Rainfall ", color="purple")
ax.set_zlabel("Yield ", color="blue")


# Parameters to create Shape/Grid using Fertilizer and Rainfall Data
x_surf = Fertilizer
y_surf = Rainfall

# The purpose of meshgrid is to create a rectangular grid out of an array of x and y values
x_surf, y_surf = np.meshgrid(x_surf, y_surf)

# To combine X and Y in order to generate Z for the surface
exog = pd.core.frame.DataFrame({
    "Fertilizer": x_surf.ravel()
    , "Rainfall": y_surf.ravel()
})

# Generating Predicted values for the shape
out = Fit2.predict(exog=exog)

# Generating the actual Shape now on Graph
ax.plot_surface(
    x_surf
    , y_surf
    , out.values.reshape(x_surf.shape)
    , rstride=1
    , cstride=1
    , color="None"
    , alpha=0.4
)
# We all know what this does , right? The End of our Assignment comes here.
plt.show()
